Empty top level xml not generating proper output for `WithNamedTypes(true)` or `WithNamedRoot(true)`

```go
{
	name:   "empty_top_level_named_type",
	xmlStr: "<a/>",
	options: []xmlstruct.GeneratorOption{
		xmlstruct.WithNamedTypes(true),
	},
	expectedStr: joinLines(
		xmlstruct.DefaultHeader,
		``,
		`package main`,
		``,
		`type A struct{}`,
	),
},
{
	name:   "empty_top_level_named_root_type",
	xmlStr: "<a/>",
	options: []xmlstruct.GeneratorOption{
		xmlstruct.WithNamedRoot(true),
	},
	expectedStr: joinLines(
		xmlstruct.DefaultHeader,
		``,
		`package main`,
		``,
		`import "encoding/xml"`,
		``,
		`type A struct {`,
		"\tXMLName xml.Name `xml:\"a\"`",
		`}`,
	),
},
```

```
--- FAIL: TestGenerator (0.00s)
    --- FAIL: TestGenerator/empty_top_level_named_type (0.00s)
        generator_test.go:607: Expected values to be equal:
             // This file is automatically generated. DO NOT EDIT.

             package main
            -
            -type A struct{}
    --- FAIL: TestGenerator/empty_top_level_named_root_type (0.00s)
        generator_test.go:607: Expected values to be equal:

             import "encoding/xml"

            -type A struct {
            -   XMLName xml.Name `xml:"a"`
            -}
            +type A struct{}
FAIL
FAIL    github.com/twpayne/go-xmlstruct 0.168s
FAIL
```

--- a/element.go
+++ b/element.go
@@ -136,7 +136,7 @@ FOR:
 
 // writeGoType writes e's Go type to w.
 func (e *element) writeGoType(w io.Writer, options *generateOptions, indentPrefix string) error {
-	if len(e.attrValues) == 0 && len(e.childElements) == 0 {
+	if len(e.attrValues) == 0 && len(e.childElements) == 0 && (!e.root || !options.namedRoot) {
 		fmt.Fprintf(w, "%s", e.charDataValue.goType(options))
 		return nil
 	}

--- a/generator.go
+++ b/generator.go
@@ -263,7 +263,7 @@ func (g *Generator) Generate() ([]byte, error) {
 		options.namedTypes = maps.Clone(g.typeElements)
 		options.simpleTypes = make(map[xml.Name]struct{})
 		for name, element := range options.namedTypes {
-			if len(element.attrValues) != 0 || len(element.childElements) != 0 {
+			if len(element.attrValues) != 0 || len(element.childElements) != 0 || element.root {
 				continue
 			}
 			options.simpleTypes[name] = struct{}{}
@@ -416,7 +416,7 @@ func (g *Generator) ObserveReader(r io.Reader) error {
 	if g.modifyDecoderFunc != nil {
 		g.modifyDecoderFunc(decoder)
 	}
-	var rootElement *element
+	var foundRootElement bool
 FOR:
 	for {
 		var token xml.Token
@@ -433,6 +433,11 @@ FOR:
 			return err
 		default:
 			if startElement, ok := token.(xml.StartElement); ok {
+				var root bool
+				if !foundRootElement {
+					foundRootElement = true
+					root = true
+				}
 				name := g.nameFunc(startElement.Name)
 				if name == (xml.Name{}) {
 					continue FOR
@@ -440,14 +445,11 @@ FOR:
 				typeElement, ok := g.typeElements[name]
 				if !ok {
 					typeElement = newElement(name)
+					typeElement.root = root
 					g.typeElements[name] = typeElement
 				}
 				if _, ok := g.typeOrder[name]; !ok {
 					g.typeOrder[name] = options.getOrder()
-				}
-				if rootElement == nil {
-					rootElement = typeElement
-					rootElement.root = true
 				}
 				if err := typeElement.observeChildElement(decoder, startElement, 0, &options); err != nil {
 					return err

--- a/generator_test.go
+++ b/generator_test.go
@@ -555,6 +555,38 @@ func TestGenerator(t *testing.T) {
 				`}`,
 			),
 		},
+		{
+			name:   "empty_top_level_named_type",
+			xmlStr: "<a/>",
+			options: []xmlstruct.GeneratorOption{
+				xmlstruct.WithNamedTypes(true),
+			},
+			expectedStr: joinLines(
+				xmlstruct.DefaultHeader,
+				``,
+				`package main`,
+				``,
+				`type A struct{}`,
+			),
+		},
+		{
+			name:   "empty_top_level_named_root_type",
+			xmlStr: "<a/>",
+			options: []xmlstruct.GeneratorOption{
+				xmlstruct.WithNamedRoot(true),
+			},
+			expectedStr: joinLines(
+				xmlstruct.DefaultHeader,
+				``,
+				`package main`,
+				``,
+				`import "encoding/xml"`,
+				``,
+				`type A struct {`,
+				"\tXMLName xml.Name `xml:\"a\"`",
+				`}`,
+			),
+		},
 	} {
 		tc := tc
 		t.Run(tc.name, func(t *testing.T) {

